src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -57,396 +56,56 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
-    // List of tokens to add/remove from the existing list
-    List<String> tokenNames = tokenMut.getTokenNameList();
-    // The existing list of tokens in the node
-    List<String> tokenList = node.tokenList();
-    TokenMutation.Type tokenMutType = tokenMut.getType();
-    if (tokenMutType == TokenMutation.Type.ADD_TOKEN) {
-      tokenList.addAll(tokenNames);
-    } else if (tokenMutType == TokenMutation.Type.DELETE_TOKEN) {
-      tokenList.removeAll(tokenNames);
-    } else {
-      // unrecognized mutation
-      return false;
-    }
-    return true;
-  }
-
-  /**
-   * Given an input graph, roots (as strings), a node map (from string to Graph Nodes), and a
-   * maximum depth, the function returns a graph with only nodes up to a max depth. Edges that don't
-   * contain nodes both reachable up until the max depth are discarded.
-   *
-   * @param graphInput the input graph, as a Mutatable Graph
-   * @param roots the name (string) of the roots
-   * @param graphNodesMap a mapping of strings to GraphNodes
-   * @param maxDepth the maximum depth of a node from a root
-   * @return a graph with nodes only a certain distance from a root
-   */
-  public MutableGraph<GraphNode> getGraphWithMaxDepth2(
-      MutableGraph<GraphNode> graphInput,
-      Set<String> roots,
-      HashMap<String, GraphNode> graphNodesMap,
-      int maxDepth) {
-
-    MutableGraph<GraphNode> graphToReturn = GraphBuilder.directed().build();
-    if (maxDepth < 0) {
-      return graphToReturn; // If max depth below 0, then return an emtpy graph
-    }
-    ArrayDeque<GraphNode> queue = new ArrayDeque<GraphNode>();
-    Map<GraphNode, Boolean> visited = new HashMap<>();
 
-    for (String rootName : roots) {
-      // Get the GraphNode object corresponding to the root name, add to the queue
-      GraphNode rootNode = graphNodesMap.get(rootName);
-      queue.add(rootNode);
-    }
-    int currentDepth = 0;
-    int currDepthElementCount = roots.size(); // Number of elements in current layer/depth
-    int nextDepthElementCount = 0; // Number of elements in the next layer/depth
-
-    while (!queue.isEmpty()) {
-      GraphNode curr =
-          queue.poll(); // Add node first, worry about edges after we have all the nodes we need
-
-      // Add to the graph to return, within the max depth height from root
-      if (!visited.containsKey(curr)) {
-        graphToReturn.addNode(curr);
-        visited.put(curr, true);
-
-        // The number of outgoing edges from the current node, number of nodes in the
-        // next layer
-        for (GraphNode gn : graphInput.successors(curr)) {
-          if (!visited.containsKey(gn)) {
-            queue.add(gn);
-            nextDepthElementCount++;
-          }
+    MutableGraph<GraphNode> graph = currDataGraph.getGraph();
+    HashMap<String, GraphNode> graphNodesMap = currDataGraph.getGraphNodesMap();
+    HashSet<String> roots = currDataGraph.getRoots();
```

Comment:
getRoots() returns a copy of the roots? Then this means that if the mutation changes a root, it won't actually be applied to the graph itself, only to the copied `roots` list, so after applying the mutation, if you call `currDataGraph.getRoots()` the roots will be outdated. 

I think currDataGraph should always be reflecting the correct state of information (graphNodes, roots) so you should probably not create a copy on the getters. 

This can also be solved by moving the mutateGraph function into DataGraph, which you are doing in a followup PR, right?

Reply:
Yup and we are also returning the roots directly with the autovalue methods.